### ISSUE TITLE
Failed to translate if build by aarch64-linux-gnu-gcc

### ISSUE THREAD

**Issue by joeyjiaojg** [revng/revng#297]
# If the code is built by aarch64-linux-gnu-gcc, the binary can't be translated succeed.
```
$ aarch64-linux-gnu-gcc -static -o crash test-revng/crash.c
$ cat test-revng/crash.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

int main()
{
  char buf[5];

  memset(buf, 0, 5);

  if (!read(0, buf, 5))
    printf("read error\n");

  if (buf[0] == 'c')
    if (buf[1] == 'r')
      if (buf[2] == 'a')
        if (buf[3] == 's')
          if (buf[4] == 'h')
            abort();

  return 0;
}
$ aarch64-linux-gnu-gcc --version
aarch64-linux-gnu-gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```
$ revng translate crash
/orchestra/root/x86_64-pc-linux-gnu/x86_64-pc-linux-gnu/binutils-bin/2.39/ld.bfd: /tmp/revng-object.o-f5648b..o: in function `helper_access_check_cp_reg':
/builds/gitlab/revng/orchestra/orchestra/sources/qemu/target-arm/op_helper.c:448: undefined reference to `g_assertion_message_expr'
Check failed at /builds/gitlab/revng/orchestra/orchestra/sources/revng/lib/Recompile/LinkForTranslation.cpp:113

ExitCode == 0
PLEASE submit a bug report to https://github.com/revng/revng and include the crash backtrace
Stack dump:
0.      Program arguments: ../root/libexec/revng/revng-pipeline -load /orchestra/root/lib64/revng/analyses/librevngBasicAnalyses.so -load /orchestra/root/lib64/revng/analyses/librevngABI.so -loa
d /orchestra/root/lib64/revng/analyses/librevngYieldPipes.so -load /orchestra/root/lib64/revng/analyses/librevngModel.so -load /orchestra/root/lib64/revng/analyses/librevngModelI
mporterBinary.so -load /orchestra/root/lib64/revng/analyses/librevngEarlyFunctionAnalysis.so -load /orchestra/root/lib64/revng/analyses/librevngABIAnalyses.so -load /orchestra/ro
ot/lib64/revng/analyses/librevngModelImporter.so -load /orchestra/root/lib64/revng/analyses/librevngTypeShrinking.so -load /orchestra/root/lib64/revng/analyses/librevngFunctionIsolation.so -load
 /orchestra/root/lib64/revng/analyses/librevngRecompile.so -P /orchestra/root/share/revng/pipelines/isolate-translate.yml -P /orchestra/root/share/revng/pipelines/yield-cfg.yml -
P /orchestra/root/share/revng/pipelines/yield-call-graph.yml -P /orchestra/root/share/revng/pipelines/yield-assembly.yml -P /orchestra/root/share/revng/pipelines/translate.yml -P
 /orchestra/root/share/revng/pipelines/enforce-abi.yml -P /orchestra/root/share/revng/pipelines/analyses-lists.yml --analyze=Import/ImportBinary/input/:Binary --analyze=Import/AddPrimitiveTypes/
 --analyze=Lift/DetectABI/module.ll/:Root --produce=Recompile/output/:Translated -i crash:begin/input -o crash.translated:Recompile/output --compile-opt-level=0
 #0 0x00007f7e9b49da5e llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /builds/gitlab/revng/orchestra/orchestra/sources/llvm-project/llvm/lib/Support/Unix/Signals.inc:570:13
 #1 0x00007f7e9b49bdc4 llvm::sys::RunSignalHandlers() /builds/gitlab/revng/orchestra/orchestra/sources/llvm-project/llvm/lib/Support/Signals.cpp:105:18
 #2 0x00007f7e9b49dfcd SignalHandler(int) /builds/gitlab/revng/orchestra/orchestra/sources/llvm-project/llvm/lib/Support/Unix/Signals.inc:415:1
 #3 0x00007f7e9aed7520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f7e9af2ba7c __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007f7e9af2ba7c __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007f7e9af2ba7c pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007f7e9aed7476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007f7e9aebd7f3 abort ./stdlib/abort.c:81:7
 #9 0x00007f7e9e8cf529 /builds/gitlab/revng/orchestra/orchestra/sources/revng/lib/Support/Assert.cpp:17:3
#10 0x00007f7e9e8cf570 revng_do_abort /builds/gitlab/revng/orchestra/orchestra/sources/revng/lib/Support/Assert.cpp:46:0
#11 0x00007f7e9926b20e TemporaryFile::cantFail(std::__1::error_code) /builds/gitlab/revng/orchestra/orchestra/sources/revng/include/revng/Support/TemporaryFile.h:43:46
#12 0x00007f7e9926b20e TemporaryFile::~TemporaryFile() /builds/gitlab/revng/orchestra/orchestra/sources/revng/include/revng/Support/TemporaryFile.h:30:22
#13 0x00007f7e9926b20e std::__1::default_delete<TemporaryFile>::operator()[abi:v160001](TemporaryFile*) const /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/v1/__memory/unique_ptr.h:6
5:5
#14 0x00007f7e9926b20e std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>::reset[abi:v160001](TemporaryFile*) /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/
v1/__memory/unique_ptr.h:297:7
#15 0x00007f7e9926b20e std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>::~unique_ptr[abi:v160001]() /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/v1/__mem
ory/unique_ptr.h:263:75
#16 0x00007f7e9926b20e void std::__1::__destroy_at[abi:v160001]<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>, 0>(std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>*) /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/v1/__memory/construct_at.h:66:13
#17 0x00007f7e9926b20e void std::__1::destroy_at[abi:v160001]<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>, 0>(std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>*) /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/v1/__memory/construct_at.h:101:5
#18 0x00007f7e9926b20e void std::__1::allocator_traits<std::__1::allocator<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>>>::destroy[abi:v160001]<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>, void, void>(std::__1::allocator<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>>&, std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>*) /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/v1/__memory/allocator_traits.h:323:9
#19 0x00007f7e9926b20e std::__1::vector<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>, std::__1::allocator<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>>>::__base_destruct_at_end[abi:v160001](std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>*) /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/v1/vector:836:9
#20 0x00007f7e9926b20e std::__1::vector<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>, std::__1::allocator<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>>>::__clear[abi:v160001]() /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/v1/vector:830:29
#21 0x00007f7e9926b20e std::__1::vector<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>, std::__1::allocator<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>>>::__destroy_vector::operator()[abi:v160001]() /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/v1/vector:446:20
#22 0x00007f7e9926b20e std::__1::vector<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>, std::__1::allocator<std::__1::unique_ptr<TemporaryFile, std::__1::default_delete<TemporaryFile>>>>::~vector[abi:v160001]() /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/clang-release/bin/../include/c++/v1/vector:456:67
#23 0x00007f7e9926b20e CommandList::~CommandList() /builds/gitlab/revng/orchestra/orchestra/sources/revng/lib/Recompile/LinkForTranslation.cpp:94:7
#24 0x00007f7e9926b20e linkForTranslation(model::Binary const&, llvm::StringRef, llvm::StringRef, llvm::StringRef) /builds/gitlab/revng/orchestra/orchestra/sources/revng/lib/Recompile/LinkForTranslation.cpp:327:1
#25 0x00007f7e9924e1ad revng::pipes::LinkForTranslation::run(pipeline::Context const&, revng::pipes::FileContainer<&revng::kinds::Binary, &revng::pipes::BinaryFileName.<char const at offset 0>, &revng::pipes::BinaryFileMIMEType.<char const at offset 0>, &revng::pipes::BinaryFileSuffix.<char const at offset 0>>&, revng::pipes::FileContainer<&revng::kinds::Object, &revng::pipes::ObjectFileName.<char const at offset 0>, &revng::pipes::ObjectFileMIMEType.<char const at offset 0>, &revng::pipes::ObjectFileSuffix.<char const at offset 0>>&, revng::pipes::FileContainer<&revng::kinds::Translated, &revng::pipes::TranslatedFileName.<char const at offset 0>, &revng::pipes::TranslatedFileMIMEType.<char const at offset 0>, &revng::pipes::TranslatedFileSuffix.<char const at offset 0>>&) /builds/gitlab/revng/orchestra/orchestra/sources/revng/lib/Recompile/LinkForTranslationPipe.cpp:32:1
#26 0x00007f7e9926932a llvm::Error::setChecked(bool) /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/llvm/include/llvm/Support/Error.h:300:13
#27 0x00007f7e9926932a llvm::Error::operator=(llvm::Error&&) /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/llvm/include/llvm/Support/Error.h:215:5
#28 0x00007f7e9926932a llvm::Error::Error(llvm::Error&&) /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/llvm/include/llvm/Support/Error.h:192:11
#29 0x00007f7e9926932a pipeline::InvokableWrapperImpl<revng::pipes::LinkForTranslation>::run(pipeline::Context&, pipeline::ContainerSet&, llvm::StringMap<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, llvm::MallocAllocator> const&) /builds/gitlab/revng/orchestra/orchestra/sources/revng/include/revng/Pipeline/Invokable.h:413:12
#30 0x00007f7e9926932a pipeline::detail::PipeWrapperImpl<revng::pipes::LinkForTranslation>::run(pipeline::Context&, pipeline::ContainerSet&, llvm::StringMap<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, llvm::MallocAllocator> const&) /builds/gitlab/revng/orchestra/orchestra/sources/revng/include/revng/Pipeline/Pipe.h:179:22
#31 0x00007f7e9ec14875 llvm::Error::getPtr() const /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/llvm/include/llvm/Support/Error.h:272:42
#32 0x00007f7e9ec14875 llvm::Error::operator bool() /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/llvm/include/llvm/Support/Error.h:235:16
#33 0x00007f7e9ec14875 llvm::cantFail(llvm::Error, char const*) /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/llvm/include/llvm/Support/Error.h:746:7
#34 0x00007f7e9ec14875 pipeline::Step::cloneAndRun(pipeline::Context&, pipeline::ContainerSet&&) /builds/gitlab/revng/orchestra/orchestra/sources/revng/lib/Pipeline/Step.cpp:94:5
#35 0x00007f7e9ec0a803 pipeline::Runner::run(llvm::StringRef, pipeline::ContainerToTargetsMap const&) /builds/gitlab/revng/orchestra/orchestra/sources/revng/lib/Pipeline/Runner.cpp:334:5
#36 0x000055df2e4dd8ca llvm::Error::getPtr() const /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/llvm/include/llvm/Support/Error.h:272:42
#37 0x000055df2e4dd8ca llvm::Error::operator bool() /builds/gitlab/revng/orchestra/orchestra/root/lib64/llvm/llvm/include/llvm/Support/Error.h:235:16
#38 0x000055df2e4dd8ca pipeline::Runner::run(llvm::StringMap<pipeline::ContainerToTargetsMap, llvm::MallocAllocator> const&) /builds/gitlab/revng/orchestra/orchestra/sources/revng/include/revng/Pipeline/Runner.h:154:16
#39 0x000055df2e4dd8ca runPipeline(pipeline::Runner&) /builds/gitlab/revng/orchestra/orchestra/sources/revng/tools/pipeline/Main.cpp:151:27
#40 0x000055df2e4dd8ca main /builds/gitlab/revng/orchestra/orchestra/sources/revng/tools/pipeline/Main.cpp:198:3
#41 0x00007f7e9aebed90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#42 0x00007f7e9aebee40 call_init ./csu/../csu/libc-start.c:128:20
#43 0x00007f7e9aebee40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#44 0x000055df2e4dc529 _start (../root/libexec/revng/revng-pipeline+0x12529)
```

# If built by aarch64-unknow-linux-musl-gcc, there is no issue in translating, but failed to execute.
```
$ aarch64-unknown-linux-musl-gcc -o crash.musl -static crash.c
$ echo hi | ./crash.musl.translated
$ echo $?
0
$ echo crash | ./crash.musl.translated
qemu: Unsupported syscall: 135
qemu: Unsupported syscall: 135
qemu: Unsupported syscall: 135
Segmentation fault
$ echo $?
139

strace data:
27095 read(0, "crash", 5)               = 5
27095 write(2, "qemu: Unsupported syscall: 135\n", 31) = 31
27095 gettid()                          = 27095
27095 tkill(27095, 0)                   = 0
27095 write(2, "qemu: Unsupported syscall: 135\n", 31) = 31
27095 write(2, "qemu: Unsupported syscall: 135\n", 31) = 31
27095 --- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=NULL} ---
27095 --- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=NULL} ---

```

# example hello.c works
```
$ cat hello.c
#include <stdio.h>

int main(int argc, char *argv[]) {
  printf("Hello world!\n");
  return 0;
}

$ revng translate hello.musl
$ ./hello.musl.translated
Hello, world!
```

**Comment by aleclearmind** [comment]
Can you attach the exact binary that's failing? Thanks.

Please consider that static binary translation has not been a core goal for a long time.
We fix bugs, but we don't expect great improvements in that sense.
We're focused on decompilation.

**Comment by joeyjiaojg** [comment]
[crash.tar.gz](https://github.com/revng/revng/files/11562341/crash.tar.gz)

attached

**Comment by aleclearmind** [comment]
Fixed in 448ebfc. Please test and reopen if this still happens.
